File: rs/core/tests/algorithm/centroid_first_interior_point_sweep_test.rs
//! Sweeps every upstream fixture through `centroid_first_interior_point` and
//! asserts two things per geometry: the result verifies, and its verification
//! outcome is the same as the outcome of the point `interior_point` would have
//! returned.
//!
//! The last of those is the real invariant. Swapping the function changes which
//! point comes back for 199 of these geometries and changes the verdict on none
//! of them.
//!
//! This reaches the crate through its public API alone, so unlike the world test
//! it can live here rather than under `src/test/`.

// The shared parser also carries each case's expected point, which this sweep
// has no use for: it compares two computed points against each other rather
// than against a fixture value.
#[allow(dead_code)]
#[path = "../utils/mod.rs"]
mod utils;

use std::fs;
use std::str::FromStr;

use geo_types::Geometry;
use interior_point::{centroid_first_interior_point, interior_point, verify_interior_point};
use utils::xml_test_parser::parse_xml_test_cases;
use wkt::Wkt;

/// How one fixture came out: how many geometries were swept, and how many of
/// them the centroid branch answered differently from `interior_point`.
#[derive(Default, PartialEq, Eq, Debug)]
struct Sweep {
    swept: usize,
    differs: usize,
}

/// Asserts one geometry and folds it into the running counts. Returns a failure
/// line instead of panicking, so one run reports every bad geometry at once.
fn sweep_one(label: &str, geometry: &Geometry<f64>, counts: &mut Sweep) -> Option<String> {
    counts.swept += 1;
    let result = centroid_first_interior_point(geometry);
    let plain = interior_point(geometry);
    if result != plain {
        counts.differs += 1;
    }
    if result.is_none() {
        // An empty input, where both functions answer None. Nothing to verify.
        return (plain.is_some()).then(|| format!("{label}: empty where interior_point was not"));
    }
    let verification = verify_interior_point(result, Some(geometry));
    if !verification.is_verified() {
        return Some(format!("{label}: result is {verification}"));
    }
    let plain_verification = verify_interior_point(plain, Some(geometry));
    if verification != plain_verification {
        return Some(format!(
            "{label}: result is {verification} where the algorithm's own point is {plain_verification}"
        ));
    }
    None
}

fn report(failures: Vec<String>, label: &str) {
    assert!(
        failures.is_empty(),
        "{} geometries in {label} failed:\n{}",
        failures.len(),
        failures.join("\n"),
    );
}

fn sweep_xml(path: &str, op: &str, expected_cases: usize, expected: Sweep) {
    let cases = parse_xml_test_cases(path, op);
    assert_eq!(cases.len(), expected_cases, "{path}: unexpected case count");
    let mut counts = Sweep::default();
    let mut failures = Vec::new();
    for case in &cases {
        let Some(geometry) = case.input.as_ref() else {
            continue;
        };
        if let Some(message) = sweep_one(&case.desc, geometry, &mut counts) {
            failures.push(message);
        }
    }
    report(failures, path);
    assert_eq!(counts, expected, "{path}: unexpected branch counts");
}

/// Not every fixture case carries an input this parser can build, which is why
/// `swept` is below the case count for both XML files.
#[test]
fn sweeps_every_interior_point_fixture_case() {
    sweep_xml(
        "../../upstream/jts/resources/testxml/general/TestInteriorPoint.xml",
        "getInteriorPoint",
        24,
        Sweep {
            swept: 21,
            differs: 3,
        },
    );
}

#[test]
fn sweeps_every_centroid_fixture_case() {
    sweep_xml(
        "../../upstream/jts/resources/testxml/general/TestCentroid.xml",
        "getCentroid",
        38,
        Sweep {
            swept: 36,
            differs: 2,
        },
    );
}

/// Splits a multi-line WKT file into individual WKT strings. Each geometry
/// starts with a type keyword at the beginning of a line.
fn split_wkt_geometries(data: &str) -> Vec<String> {
    let mut geometries = Vec::new();
    let mut current = String::new();
    for line in data.lines() {
        let trimmed = line.trim();
        if trimmed.is_empty() {
            continue;
        }
        let starts_new = [
            "POINT",
            "LINESTRING",
            "POLYGON",
            "MULTIPOINT",
            "MULTILINESTRING",
            "MULTIPOLYGON",
            "GEOMETRYCOLLECTION",
        ]
        .iter()
        .any(|keyword| trimmed.starts_with(keyword));
        if starts_new && !current.is_empty() {
            geometries.push(current.trim().to_string());
            current = String::new();
        }
        if !current.is_empty() {
            current.push(' ');
        }
        current.push_str(trimmed);
    }
    if !current.trim().is_empty() {
        geometries.push(current.trim().to_string());
    }
    geometries
}

#[test]
fn sweeps_every_world_geometry() {
    let data = fs::read_to_string("../../upstream/jts/resources/testdata/world.wkt")
        .expect("Failed to read world.wkt — run cargo from rs/");
    let wkt_strings = split_wkt_geometries(&data);
    assert_eq!(
        wkt_strings.len(),
        244,
        "unexpected world.wkt geometry count"
    );

    let mut counts = Sweep::default();
    let mut failures = Vec::new();
    for (index, wkt_str) in wkt_strings.iter().enumerate() {
        let parsed = Wkt::from_str(wkt_str).expect("every world.wkt geometry parses");
        let geometry = Geometry::<f64>::try_from(parsed).expect("and converts");
        if let Some(message) = sweep_one(&format!("geometry {}", index + 1), &geometry, &mut counts)
        {
            failures.push(message);
        }
    }
    report(failures, "world.wkt");
    // Every one of the 244 is areal, so none is skipped, and the centroid is
    // accepted often enough that the fallback fires for only 50 of them.
    assert_eq!(
        counts,
        Sweep {
            swept: 244,
            differs: 194,
        }
    );
}

File: js/test/CentroidFirstInteriorPointSweepTest.ts
/**
 * Centroid-first sweep over the three upstream fixtures.
 *
 * The invariant is that swapping `interiorPoint` for `centroidFirstInteriorPoint`
 * changes which point comes back but never how good it is: for every fixture
 * geometry the two agree on whether a point exists at all, and they verify to the
 * same outcome. The branch counts are asserted too, so a change that quietly
 * stopped accepting centroids would fail here rather than pass unnoticed.
 *
 * There is no `@jts` anchor on this file: this surface has no JTS counterpart,
 * neither in the library nor in its tests.
 */
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import { resolve } from "node:path";
import type { Geometry } from "geojson";
import { centroidFirstInteriorPoint } from "../src/CentroidFirstInteriorPoint.ts";
import { getCentroid } from "../src/algorithm/Centroid.ts";
import { dimensionNonEmpty, interiorPoint } from "../src/algorithm/InteriorPoint.ts";
import { Verification, verifyInteriorPoint } from "../src/VerifyInteriorPoint.ts";
import { parseXmlTestCases } from "./utils/XmlTestParser.ts";
import { parseWktFile } from "./utils/WktParser.ts";

const repoRoot = resolve(import.meta.dirname, "../..");

const interiorPointGeometries = parseXmlTestCases(
  resolve(repoRoot, "upstream/jts/resources/testxml/general/TestInteriorPoint.xml"),
  "getInteriorPoint",
).map((tc) => tc.input);

const centroidGeometries = parseXmlTestCases(
  resolve(repoRoot, "upstream/jts/resources/testxml/general/TestCentroid.xml"),
  "getCentroid",
).map((tc) => tc.input);

const worldGeometries: (Geometry | null)[] = parseWktFile(
  resolve(repoRoot, "upstream/jts/resources/testdata/world.wkt"),
);

function sweep(label: string, geometries: (Geometry | null)[], expectedCount: number): void {
  describe(`centroid-first sweep - ${label}`, () => {
    it(`loads ${expectedCount} geometries`, () => {
      assert.equal(geometries.length, expectedCount);
    });

    geometries.forEach((geometry, index) => {
      it(`geometry[${index}]: verifies exactly as interiorPoint does`, () => {
        const point = centroidFirstInteriorPoint(geometry);
        const fallback = interiorPoint(geometry);
        assert.equal(point === null, fallback === null, `${label}[${index}]: nullity differs`);
        const outcome = verifyInteriorPoint(point, geometry);
        assert.equal(outcome, verifyInteriorPoint(fallback, geometry), `${label}[${index}]: ${outcome}`);
        if (point === null) {
          assert.equal(outcome, Verification.Unverifiable);
          return;
        }
        assert.notEqual(outcome, Verification.OffGeometry, `${label}[${index}]: ${outcome}`);
      });
    });
  });
}

sweep("TestInteriorPoint.xml", interiorPointGeometries, 24);
sweep("TestCentroid.xml", centroidGeometries, 38);
sweep("world.wkt", worldGeometries, 244);

/** How many areal geometries had their centroid accepted. */
function countCentroidBranch(geometries: (Geometry | null)[]): number {
  let accepted = 0;
  for (const geometry of geometries) {
    if (geometry === null) continue;
    const point = centroidFirstInteriorPoint(geometry);
    if (point === null || dimensionNonEmpty(geometry) !== 2) continue;
    const cent = getCentroid(geometry);
    if (cent !== null && cent.length === point.length && cent.every((o, i) => o === point[i])) accepted += 1;
  }
  return accepted;
}

function countOutcomes(geometries: (Geometry | null)[]): Record<Verification, number> {
  const counts: Record<Verification, number> = {
    [Verification.Interior]: 0,
    [Verification.OnGeometry]: 0,
    [Verification.OffGeometry]: 0,
    [Verification.Unverifiable]: 0,
  };
  for (const geometry of geometries) {
    counts[verifyInteriorPoint(centroidFirstInteriorPoint(geometry), geometry)] += 1;
  }
  return counts;
}

describe("centroid-first sweep - totals", () => {
  const all = [...interiorPointGeometries, ...centroidGeometries, ...worldGeometries];

  it("reaches no off-geometry across all 306 fixture geometries", () => {
    assert.equal(all.length, 306);
    assert.deepEqual(countOutcomes(all), {
      [Verification.Interior]: 260,
      [Verification.OnGeometry]: 39,
      [Verification.OffGeometry]: 0,
      [Verification.Unverifiable]: 7,
    });
  });

  it("accepts the centroid for 207 of the 306, per fixture 5, 8 and 194", () => {
    assert.equal(countCentroidBranch(interiorPointGeometries), 5);
    assert.equal(countCentroidBranch(centroidGeometries), 8);
    assert.equal(countCentroidBranch(worldGeometries), 194);
    assert.equal(countCentroidBranch(all), 207);
  });

  it("returns a different point from interiorPoint for 199 of the 306", () => {
    let differ = 0;
    for (const geometry of all) {
      const point = centroidFirstInteriorPoint(geometry);
      const fallback = interiorPoint(geometry);
      if (point === null || fallback === null) continue;
      if (point.length !== fallback.length || !point.every((o, i) => o === fallback[i])) differ += 1;
    }
    assert.equal(differ, 199);
  });
});
